File: 模板/go/SegmentTree/queryMin.go
package Solution

type STNode struct {
	left  *STNode
	right *STNode
	val   int64
	lazy  int64
}

type SegmentTree struct {
	root *STNode
}

func NewSegmentTree() *SegmentTree {
	st := &SegmentTree{
		root: new(STNode),
	}
	return st
}

func (st *SegmentTree) assign(node *STNode, l, r, start, end int, x int64) {
	if l == start && r == end {
		node.lazy = x
		return
	}
	st.pushdown(node)
	mid := (l + r) >> 1
	if end <= mid {
		st.assign(node.left, l, mid, start, end, x)
	} else if start > mid {
		st.assign(node.right, mid+1, r, start, end, x)
	} else {
		st.assign(node.left, l, mid, start, mid, x)
		st.assign(node.right, mid+1, r, mid+1, end, x)
	}
	st.pushup(node, mid-l+1, r-mid)
}

func (st *SegmentTree) query(node *STNode, l, r, start, end int) (res int64) {
	if l == start && r == end {
		return node.val + node.lazy
	}
	st.pushdown(node)
	mid := (l + r) >> 1
	if end <= mid {
		res = st.query(node.left, l, mid, start, end)
	} else if start > mid {
		res = st.query(node.right, mid+1, r, start, end)
	} else {
		res = min(st.query(node.left, l, mid, start, mid),
			st.query(node.right, mid+1, r, mid+1, end))
	}
	st.pushup(node, mid-l+1, r-mid)
	return
}

func (st *SegmentTree) pushdown(node *STNode) {
	if node.left == nil {
		node.left = new(STNode)
	}
	if node.right == nil {
		node.right = new(STNode)
	}
	if node.lazy > 0 {
		node.left.lazy = node.lazy
		node.right.lazy = node.lazy
		node.lazy = 0
	}
}

func (st *SegmentTree) pushup(node *STNode, ln, rn int) {
	node.val = min(node.left.val+node.left.lazy*int64(ln), node.right.val+node.right.lazy*int64(rn))
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

File: 模板/python/SegmentTree/动态开点/queryMin.py
class STNode:
    def __init__(self, left=None, right=None, val=0, lazy=0):
        self.left = left
        self.right = right
        self.val = val
        self.lazy = lazy

class SegmentTree:
    def __init__(self):
        self.root = STNode()
    
    def assign(self, node, l, r, start, end, x):
        if l == start and r == end:
            node.lazy = x
            return
        mid = l+r>>1
        self.pushdown(node)
        if end <= mid:
            self.assign(node.left, l, mid, start, end, x)
        elif start > mid:
            self.assign(node.right, mid+1, r, start, end, x)
        else:
            self.assign(node.left, l, mid, start, mid, x)
            self.assign(node.right, mid+1, r, mid+1, end, x)
        self.pushup(node)

    def query(self, node, l, r, start, end):
        if l == start and r == end:
            return node.lazy + node.val
        mid = l+r>>1
        self.pushdown(node)
        if end <= mid:
            res = self.query(node.left, l, mid, start, end)
        elif start > mid:
            res = self.query(node.right, mid+1, r, start, end)
        else:
            res = min(self.query(node.left, l, mid, start, mid),
                self.query(node.right, mid+1, r, mid+1, end))
        self.pushup(node)
        return res
    
    def pushdown(self, node):
        if not node.left:
            node.left = STNode()
        if not node.right:
            node.right = STNode()
        if node.lazy:
            node.left.lazy = node.lazy
            node.right.lazy = node.lazy
            node.lazy = 0
    
    def pushup(self, node):
        node.val = min(node.left.val + node.left.lazy, node.right.val + node.right.lazy)


if __name__ == "__main__":
    n, m = [int(i) for i in input().split()]
    nums = [int(i) for i in input().split()]
    st = SegmentTree()
    for i in range(1, n+1):
        st.assign(st.root, 1, n, i, i, nums[i-1])
    for _ in range(m):
        a = [int(i) for i in input().split()]
        if a[0] == 1:
            st.assign(st.root, 1, n, a[1], a[2], a[3])
        else:
            print(st.query(st.root, 1, n, a[1], a[2]))
# 5 17
# 1 5 4 2 3
# 2 1 5
# 2 2 4
# 2 1 5
# 2 1 5
# 1 2 3 2
# 2 1 5
# 2 3 4
# 2 1 5
# 1 1 5 1
# 2 1 5
# 2 1 1
# 2 2 2
# 2 3 3
# 2 4 4
# 2 5 5
# 2 1 5
# 2 1 5

# 5 10
# 1 5 4 2 3
# 2 2 4
# 1 2 3 2
# 2 3 4
# 1 1 5 1
# 2 1 4
# 2 1 1
# 2 2 2
# 2 3 3
# 2 4 4
# 2 5 5
